code/modules/modular_computers/file_system/programs/engineering/device_editor.dm

```diff
@@ -0,0 +1,142 @@
+	program_icon_state = "generic"
+	program_key_state = "generic_key"
+	program_menu_icon = "note"
+	size = 12
+	requires_exonet = 1
+
+	nanomodule_path = /datum/nano_module/program/device_editor
+
+/datum/computer_file/program/device_editor/can_run(var/mob/living/user, var/loud = 0, var/access_to_check)
+	. = ..()
+	var/datum/extension/interactive/ntos/os = get_extension(computer, /datum/extension/interactive/ntos)
+	var/obj/item/stock_parts/computer/rfid_programmer/programmer = os && os.get_component(/obj/item/stock_parts/computer/rfid_programmer)
+	if(programmer)
+		if(istype(programmer.get_device(), /obj/item/stock_parts/exonet_lock))
+			// Need admin permissions
+			var/obj/item/card/id/exonet/I = user.GetIdCard()
+			if(!I)
+				// No valid ID...
+				if(loud)
+					to_chat(user, SPAN_NOTICE("\The [computer] flashes an \"RFID Error - Unable to scan ID\" warning."))
+				return 0
+			var/obj/item/stock_parts/exonet_lock/lock = programmer.get_device()
+			if(!lock.ennid)
+				// pffft no security.
+				return
+			var/datum/exonet/network = GLOB.exonets[lock.ennid]
+			if(!network)
+				// network doesn't exist anymore?
+				return
+			if(I.ennid != lock.ennid)
+				// Wrong ID network.
+				if(loud)
+					to_chat(user, SPAN_NOTICE("\The [computer] flashes an \"Access Denied\" warning."))
+				return 0
+			if(!(I.user_id in network.administrators))
+				// Not an administrator.
+				if(loud)
+					to_chat(user, SPAN_NOTICE("\The [computer] flashes an \"Access Denied\" warning."))
+				return 0
+
+/datum/nano_module/program/device_editor
+	name = "NTOS Device Editor"
+
+/datum/nano_module/program/device_editor/proc/get_all_grants()
+	var/list/grants = list()
+	var/datum/extension/exonet_device/exonet = get_exonet_device()
+	for(var/obj/machinery/computer/exonet/mainframe/mainframe in exonet.get_mainframes())
+		if(!mainframe.operable())
+			continue
+		for(var/datum/computer_file/data/grant_record/GR in mainframe.stored_files)
+			grants.Add(GR)
+	return grants
+
+/datum/computer_file/program/device_editor/Topic(href, href_list)
+	if(..())
+		return TOPIC_HANDLED
+
+	var/datum/nano_module/program/device_editor/NMM = NM
+	var/obj/item/stock_parts/computer/rfid_programmer/programmer = NMM.get_functional_component(/obj/item/stock_parts/computer/rfid_programmer)
+	var/obj/item/stock_parts/exonet_lock/lock = programmer.get_device()
+
+	if(href_list["PRG_refresh"])
+		error = null
+		return TOPIC_HANDLED
+	if(href_list["ennid"])
+		var/new_ennid = sanitize(input(usr, "Enter exonet ennid or leave blank to cancel:", "Change ENNID"))
+		if(!new_ennid)
+			return 1
+		lock.ennid = new_ennid
+		lock.grants = list()
+	if(href_list["key"])
+		var/new_key = sanitize(input(usr, "Enter exonet keypass or leave blank to cancel:", "Change key"))
+		if(!new_key)
+			return 1
+		lock.keydata = new_key
+		lock.grants = list()
+	if(href_list["PRG_disable"])
+		lock.tightened = FALSE
+	if(href_list["PRG_enable"])
+		lock.tightened = TRUE
```

Comment:
Not a fan of this being software controlled. If you want to restrict access to it, install a hardware lock as well.

Reply:
Not a restricted access thing. It's to prevent circumventing req_access-level control by removing the lock. Balance issue.

Followup evolve on this implementation is removing req_access entirely, which is redundant, and this needing the same-ish circumvention gameplay pathways.

Reply:
I understand how it works, I just don't like it from a design perspective. I think a pathway for hardware circumvention should exist.

Reply:
EMAG? Could add an attack by welder that shunts it into default mode, whatever the user set it to.

Reply:
There's also still wires, for airlocks.

Reply:
I think having it be removable and adding a hardware lock if you want it to not be is sufficient. @comma can weigh in if he feels differently.

Reply:
There's parts already existing for preventing hardware access, don't see why this is needed.

Reply:
What do you mean preventing hardware access? Exonet lock doesn't prevent you from accessing the hardware, only the lock itself?

Reply:
Unrelated question, if the access_lock prevents opening the access panel, how does an access_lock get the permissions to unlock/lock the access lock if the permissions to operate the device are on the exolock?

Reply:
pretty sure that means it's not preventing jack shit then because if you got to hatch-open stage you can just do full deconstruction step with a crowbar and rebuild machine without the exonet lock part.

Reply:
Ugh. That's true.